subsys/bluetooth/host/l2cap.c

```diff
@@ -582,14 +582,27 @@ int bt_l2cap_send_cb(struct bt_conn *conn, uint16_t cid, struct net_buf *buf,
 		     bt_conn_tx_cb_t cb, void *user_data)
 {
 	struct bt_l2cap_hdr *hdr;
+	struct net_buf_simple_state state;
+	int err;
 
 	BT_DBG("conn %p cid %u len %zu", conn, cid, net_buf_frags_len(buf));
 
+	net_buf_simple_save(&buf->b, &state);
+
 	hdr = net_buf_push(buf, sizeof(*hdr));
 	hdr->len = sys_cpu_to_le16(buf->len - sizeof(*hdr));
 	hdr->cid = sys_cpu_to_le16(cid);
 
-	return bt_conn_send_cb(conn, buf, cb, user_data);
+	/* Take a ref since bt_conn_send_cb takes ownership of the buffer */
+	err = bt_conn_send_cb(conn, net_buf_ref(buf), cb, user_data);
```

Comment:
Right now `buf` will not be freed in case of error. This solves one case when user expects such behavior. But there are other places where `bt_l2cap_send_cb` is called and in those places expect unrefed buffer.

Function documentation also say about unrefing:
```
/* Send L2CAP PDU over a connection
 *
 * Buffer ownership is transferred to stack so either in case of success
 * or error the buffer will be unref internally.
 *
 * Calling this from RX thread is assumed to never fail so the return can be
 * ignored.
 */
int bt_l2cap_send_cb(struct bt_conn *conn, uint16_t cid, struct net_buf *buf,
		     bt_conn_tx_cb_t cb, void *user_data);
```

Reply:
This is no longer the case, it can fail as you are experiencing, I can update the documentation if that is your problem with it.

Reply:
Documentation would be wrong, so yes, it needs update. Another thing is that other places such as `smp_br_send()` and `smp_send()` in `subsys/bluetooth/host/smp.c` expect buffer to be unreferenced.

Reply:
Yep, I guess we would need to ref it from ATT then but then the whole logic of save/restore would need to be done by in att.c.

Reply:
So in general I don't think `Bluetooth: L2CAP: Fix not restoring buffer state when send fails` commit title is valid. Documentation says about taking ownership and unreferencing buffer, so state of unrefed buffer is invalid anyway. So it is more like user of `bt_l2cap_send_cb()` is not doing the right thing, by trying to reuse the buffer. If that needs to be changed, than we need to look at whole codebase.

Reply:
Ive updated it to be just affect ATT code.